src/geometry/euclidean/basic.lean

```diff
@@ -642,4 +660,124 @@ calc dist (r1 • v +ᵥ p1) (r2 • v +ᵥ p2) * dist (r1 • v +ᵥ p1) (r2 
 ... = dist p1 p2 * dist p1 p2 + (r1 - r2) * (r1 - r2) * (∥v∥ * ∥v∥)
   : by { rw [dist_eq_norm_vsub V p1, abs_mul_abs_self, mul_assoc] }
 
+/-- Reflection in a nonempty affine subspace, whose direction is
+complete (generalization of both the common cases of reflection in a
+point and reflection in a codimension-one subspace). -/
```

Comment:
I don't understand this docstring. To me it reads like codimension 1 subspaces are always complete.

Did you try defining `reflection` (and `orthogonal_projection`) without completness and non-emptyness assumptions? It would make statements much more readable to have `reflection : affine_subspace ℝ P → P ≃ᵢ P`. The definition could be something like
```lean
def reflection {s : affine_subspace ℝ P} := 
if (h : (s : set P).nonempty ∧ is_complete (s.direction : set V)) then ... else id
```
(where `id` should be replaced to by whatever is the bundled identity isometry). Then of course a lot of lemmas would have assumptions `(hn : (s : set P).nonempty)` and `(hc : is_complete (s.direction : set V))` but you will be able to write the much more readable `reflection s p`, and you'll even have the good surprise that some lemmas don't need those assumptions (for instance `reflection_symm` and `reflection_reflection`).

The topology part of mathlib is full of this trick, and it's a general theme of formalized mathematics that you don't want to ask for permission before writing stuff, only when rewriting stuff.

Reply:
Thanks for the advice. I've filed #3952 that redefines `orthogonal_projection`, for both real inner product spaces and Euclidean affine spaces, to avoid such assumptions on the definition and use the identity map when those properties don't hold. I think that with that definition, no `if` will actually be needed in the definition of `reflection`, because the existing definition of `reflection` (minus hypotheses) is the identity map when `orthogonal_projection` is the identity map.

I've rewritten that doc string to make it clearer that it's commenting on how the word "reflection" can be used with more or less general definitions. It will need further changes once merged with the changes from #3952.

Reply:
Great! I was indeed hoping the `if` could be only in orthogonal projection since the assumptions came from there. I kicked #3952 to the bors queue.

Congratulations for yet another completed initiation rite! Switching to such a kind of definition feels super weird in the beginning, but then everything becomes a lot more fun.

Reply:
I've now merged in the changes from #3952 and updated the `reflection` code so that `nonempty` and `is_complete` are not required by the definition or by most lemmas.